File: io/InputStream.cs
using System.Diagnostics;
using System.IO;
using System.Text;

/*
 * Copyright (c) 2016-2017, Adam <Adam@sigterm.info>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
namespace OSRSCache.io
{

	public class InputStream : Stream
	{
		private static readonly char[] CHARACTERS = new char[] {'\u20ac', '\u0000', '\u201a', '\u0192', '\u201e', '\u2026', '\u2020', '\u2021', '\u02c6', '\u2030', '\u0160', '\u2039', '\u0152', '\u0000', '\u017d', '\u0000', '\u0000', '\u2018', '\u2019', '\u201c', '\u201d', '\u2022', '\u2013', '\u2014', '\u02dc', '\u2122', '\u0161', '\u203a', '\u0153', '\u0000', '\u017e', '\u0178'};

		private readonly ByteBuffer buffer;

		public InputStream(sbyte[] buffer)
		{
			this.buffer = ByteBuffer.wrap(buffer);
		}

		public virtual sbyte[] Array
		{
			get
			{
				Debug.Assert(buffer.hasArray());
				return buffer.array();
			}
		}

		public override string ToString()
		{
			return "InputStream{" + "buffer=" + buffer + '}';
		}

		public virtual int read24BitInt()
		{
			return (this.readUnsignedByte() << 16) + (this.readUnsignedByte() << 8) + this.readUnsignedByte();
		}

		public virtual void skip(int length)
		{
			int pos = buffer.position();
			pos += length;
			buffer.position(pos);
		}

		public virtual int Offset
		{
			set
			{
				buffer.position(value);
			}
			get
			{
				return buffer.position();
			}
		}


		public virtual int Length
		{
			get
			{
				return buffer.limit();
			}
		}

		public virtual int remaining()
		{
			return buffer.remaining();
		}

		public virtual sbyte readByte()
		{
			return buffer.get();
		}

		public virtual void readBytes(sbyte[] buffer, int off, int len)
		{
			this.buffer.get(buffer, off, len);
		}

		public virtual void readBytes(sbyte[] buffer)
		{
			this.buffer.get(buffer);
		}

		public virtual int readUnsignedByte()
		{
			return this.readByte() & 0xFF;
		}

		public virtual int readUnsignedShort()
		{
			return buffer.getShort() & 0xFFFF;
		}

		public virtual short readShort()
		{
			return buffer.getShort();
		}

		public virtual int readInt()
		{
			return buffer.getInt();
		}

		public virtual sbyte peek()
		{
			return buffer.get(buffer.position());
		}

		public virtual int readBigSmart()
		{
			return peek() >= 0 ? (this.readUnsignedShort() & 0xFFFF) : (this.readInt() & int.MaxValue);
		}

		public virtual int readBigSmart2()
		{
			if (peek() < 0)
			{
				return readInt() & int.MaxValue; // and off sign bit
			}
			int value = readUnsignedShort();
			return value == 32767 ? -1 : value;
		}

		public virtual int readShortSmart()
		{
			int peek = this.peek() & 0xFF;
			return peek < 128 ? this.readUnsignedByte() - 64 : this.readUnsignedShort() - 0xc000;
		}

		public virtual int readUnsignedShortSmart()
		{
			int peek = this.peek() & 0xFF;
			return peek < 128 ? this.readUnsignedByte() : this.readUnsignedShort() - 0x8000;
		}

		public virtual int readUnsignedIntSmartShortCompat()
		{
			int var1 = 0;

			int var2;
			for (var2 = this.readUnsignedShortSmart(); var2 == 32767; var2 = this.readUnsignedShortSmart())
			{
				var1 += 32767;
			}

			var1 += var2;
			return var1;
		}

		public virtual string readString()
		{
			StringBuilder sb = new StringBuilder();

			for (; ;)
			{
				int ch = this.readUnsignedByte();

				if (ch == 0)
				{
					break;
				}

				if (ch >= 128 && ch < 160)
				{
					char var7 = CHARACTERS[ch - 128];
					if ((char)0 == var7)
					{
						var7 = '?';
					}

					ch = var7;
				}

				sb.Append((char) ch);
			}
			return sb.ToString();
		}

		public virtual string readString2()
		{
			if (this.readByte() != 0)
			{
				throw new System.InvalidOperationException("Invalid jstr2");
			}
			else
			{
				return readString();
			}
		}

		public virtual string readStringOrNull()
		{
			if (this.peek() != 0)
			{
				return readString();
			}
			else
			{
				this.readByte(); // discard
				return null;
			}
		}

		public virtual int readVarInt()
		{
			sbyte var1 = this.readByte();

			int var2;
			for (var2 = 0; var1 < 0; var1 = this.readByte())
			{
				var2 = (var2 | var1 & 127) << 7;
			}

			return var2 | var1;
		}

		public virtual sbyte[] Remaining
		{
			get
			{
				sbyte[] b = new sbyte[buffer.remaining()];
				buffer.get(b);
				return b;
			}
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: @Override public int read() throws java.io.IOException
		public override int read()
		{
			return this.readUnsignedByte();
		}
	}

}

File: src/main/java/net/runelite/cache/io/InputStream.cpp
#include "InputStream.h"

namespace net::runelite::cache::io
{
	using IOException = java::io::IOException;
	using ByteBuffer = java::nio::ByteBuffer;
const std::vector<wchar_t> InputStream::CHARACTERS = std::vector<wchar_t> {L'\u20ac', L'\u0000', L'\u201a', L'\u0192', L'\u201e', L'\u2026', L'\u2020', L'\u2021', L'\u02c6', L'\u2030', L'\u0160', L'\u2039', L'\u0152', L'\u0000', L'\u017d', L'\u0000', L'\u0000', L'\u2018', L'\u2019', L'\u201c', L'\u201d', L'\u2022', L'\u2013', L'\u2014', L'\u02dc', L'\u2122', L'\u0161', L'\u203a', L'\u0153', L'\u0000', L'\u017e', L'\u0178'};

	InputStream::InputStream(std::vector<signed char> &buffer) : buffer(ByteBuffer::wrap(buffer))
	{
	}

	std::vector<signed char> InputStream::getArray()
	{
		assert(buffer->hasArray());
		return buffer->array();
	}

	std::wstring InputStream::toString()
	{
		return L"InputStream{" + L"buffer=" + buffer + L'}';
	}

	int InputStream::read24BitInt()
	{
		return (this->readUnsignedByte() << 16) + (this->readUnsignedByte() << 8) + this->readUnsignedByte();
	}

	void InputStream::skip(int length)
	{
		int pos = buffer->position();
		pos += length;
		buffer->position(pos);
	}

	void InputStream::setOffset(int offset)
	{
		buffer->position(offset);
	}

	int InputStream::getOffset()
	{
		return buffer->position();
	}

	int InputStream::getLength()
	{
		return buffer->limit();
	}

	int InputStream::remaining()
	{
		return buffer->remaining();
	}

	signed char InputStream::readByte()
	{
		return buffer->get();
	}

	void InputStream::readBytes(std::vector<signed char> &buffer, int off, int len)
	{
		this->buffer->get(buffer, off, len);
	}

	void InputStream::readBytes(std::vector<signed char> &buffer)
	{
		this->buffer->get(buffer);
	}

	int InputStream::readUnsignedByte()
	{
		return this->readByte() & 0xFF;
	}

	int InputStream::readUnsignedShort()
	{
		return buffer->getShort() & 0xFFFF;
	}

	short InputStream::readShort()
	{
		return buffer->getShort();
	}

	int InputStream::readInt()
	{
		return buffer->getInt();
	}

	long long InputStream::readLong()
	{
		return buffer->getLong();
	}

	signed char InputStream::peek()
	{
		return buffer->get(buffer->position());
	}

	int InputStream::readBigSmart()
	{
		return peek() >= 0 ? (this->readUnsignedShort() & 0xFFFF) : (this->readInt() & std::numeric_limits<int>::max());
	}

	int InputStream::readBigSmart2()
	{
		if (peek() < 0)
		{
			return readInt() & std::numeric_limits<int>::max(); // and off sign bit
		}
		int value = readUnsignedShort();
		return value == 32767 ? -1 : value;
	}

	int InputStream::readShortSmart()
	{
		int peek = this->peek() & 0xFF;
		return peek < 128 ? this->readUnsignedByte() - 64 : this->readUnsignedShort() - 0xc000;
	}

	int InputStream::readUnsignedShortSmart()
	{
		int peek = this->peek() & 0xFF;
		return peek < 128 ? this->readUnsignedByte() : this->readUnsignedShort() - 0x8000;
	}

	int InputStream::readUnsignedIntSmartShortCompat()
	{
		int var1 = 0;

		int var2;
		for (var2 = this->readUnsignedShortSmart(); var2 == 32767; var2 = this->readUnsignedShortSmart())
		{
			var1 += 32767;
		}

		var1 += var2;
		return var1;
	}

	std::wstring InputStream::readString()
	{
		std::shared_ptr<StringBuilder> sb = std::make_shared<StringBuilder>();

		for (; ;)
		{
			int ch = this->readUnsignedByte();

			if (ch == 0)
			{
				break;
			}

			if (ch >= 128 && ch < 160)
			{
				wchar_t var7 = CHARACTERS[ch - 128];
				if (0 == var7)
				{
					var7 = L'?';
				}

				ch = var7;
			}

			sb->append(static_cast<wchar_t>(ch));
		}
		return sb->toString();
	}

	std::wstring InputStream::readString2()
	{
		if (this->readByte() != 0)
		{
			throw IllegalStateException("Invalid jstr2");
		}
		else
		{
			return readString();
		}
	}

	std::wstring InputStream::readStringOrNull()
	{
		if (this->peek() != 0)
		{
			return readString();
		}
		else
		{
			this->readByte(); // discard
			return L"";
		}
	}

	int InputStream::readVarInt()
	{
		signed char var1 = this->readByte();

		int var2;
		for (var2 = 0; var1 < 0; var1 = this->readByte())
		{
			var2 = (var2 | var1 & 127) << 7;
		}

		return var2 | var1;
	}

	int InputStream::readVarInt2()
	{
		int value = 0;
		int bits = 0;
		int read;
		do
		{
			read = readUnsignedByte();
			value |= (read & 0x7F) << bits;
			bits += 7;
		} while (read > 127);
		return value;
	}

	std::vector<signed char> InputStream::getRemaining()
	{
		std::vector<signed char> b(buffer->remaining());
		buffer->get(b);
		return b;
	}

	int InputStream::read()
	{
		return this->readUnsignedByte();
	}
}
